pkg/controller/node/node_controller.go

```diff
@@ -279,10 +282,37 @@ func (ctrl *Controller) getPoolForNode(node *corev1.Node) (*mcfgv1.MachineConfig
 		// This is not an error, as there might be nodes in cluster that are not managed by machineconfigpool.
 		return nil, nil
 	}
-	if len(pools) > 1 {
-		return nil, fmt.Errorf("node %s belongs to more than one MachineConfigPool, cannot proceed with this Node", node.Name)
+
+	var master, worker *mcfgv1.MachineConfigPool
+	var custom []*mcfgv1.MachineConfigPool
+	for _, pool := range pools {
+		if pool.Name == "master" {
+			master = pool
+		} else if pool.Name == "worker" {
+			worker = pool
+		} else {
+			custom = append(custom, pool)
+		}
 	}
-	return pools[0], nil
+
+	if len(custom) > 1 {
+		return nil, fmt.Errorf("node %s belongs to %d custom roles, cannot proceed with this Node", node.Name, len(custom))
+	} else if len(custom) == 1 {
+		// We don't support making custom pools for masters
+		if master != nil {
+			return nil, fmt.Errorf("node %s has both master role and custom role %s", node.Name, custom[0].Name)
+		}
+		// One custom role, let's use its pool
+		return custom[0], nil
+	} else if master != nil {
```

Comment:
just to check if i am misreading but to get into this third else if.. won't we also get into it if a node is only tagged master and not worker? so shouldnt we check: master != nil and worker == nil??

or am i misreading?

Reply:
> won't we also get into it if a node is only tagged master and not worker?

If it's tagged master and not worker, it's master right?

> so shouldnt we check: master != nil and worker == nil??

We're explicitly choosing master in that case for the CRC and single-machine-test cases; if master is set we want master, regardless of whether `worker` is nil-or-not.

Reply:
thanks i needed that! 

if 1 custom: custom
if master: master
if only worker: worker

Got it!